vignettes/design_principle.Rmd

```diff
@@ -0,0 +1,90 @@
+
+
+
+There are several R packages designed to access data from specific HIS. They are usually tied to a single HIS and can't be used to query others. {fingertipsR}, {REDCapR}, {godataR}, {globaldothealth} can be used to fetch data from [Fingertips](https://fingertips.phe.org.uk/), [REDCap](https://projectredcap.org/software/), [goData](https://www.who.int/tools/godata), and [Global.Health](https://global.health/) respectively. In these packages, the API end points are accessed differently. This introduces the challenge of having a unified framework for importing data from these HIS into R.
+
+To address this challenge, PIs of the [Epiverse-TRACE](https://data.org/initiatives/epiverse/) came with a list of potential data sources for which we aim at building a tool to make requests and fetch the data of interest in the same way. The data sources include: distributed health information systems, and public databases as shown in the figure below.
+
+![readepi roadmap](../man/figures/roadmap-readepi.drawio.png)
+
+## Scope
+
+The {readepi} package intends to import data from two common sources of institutional health-related data: health information systems (HIS) wrapped with specific program interfaces (APIs) and relational database management systems (RDBMS) that run on specific servers. Each source has its own import function, `read_from_api()` for HIS and `read_from_server()` for RDBMS. There is also an auxiliary function, `get_metadata()`, that retrieves metadata from a specified API and offers interoperability with the [{cleanepi} R package](https://epiverse-trace.github.io/cleanepi/).     
+The current version of {readepi} supports importing data from HIS APIs such as REDCap (Research Electronic Data Capture), DHIS2 (District Health Information System 2), Fingertips, GoData, and Globaldothealth, as well as RDBMS such as MS SQL, MS ACCESS, SQlite, MYSQL, and PostgreSQL. 
+
+![readepi design](../man/figures/readepi_design.svg)
+
+
+## Output
+
+The two import functions return a `list` object containing one or more `data frames`. Each `data frame` corresponds to the data from a specified table or file. The `get_metadata()` function returns a data dictionary containing information about the data structure. 
+
+## Design decisions
+
+The aim of {readepi} is to simplify and standardize the process of fetching health data from APIs and servers. We ambition to make it requires minimal information to access and pull the data of interest from the target source.  
+
+1. The `read_from_api()` function takes only two arguments to import data from an API: an object of class **credential** and an object of class **api**.
+
+The **credential** object is a list-like object that stores the information required to access the API. Its main attributes are:
```

Comment:
Did you consider the benefits & downsides of list vs named vector?

Or should it be an actual class?

Reply:
Is there value of pulling the credential object outside of the `read_from_()` call?

The workflow would be:

```r
authenticate(name, credentials)
read()
read()
read()
```

Reply:
I opt for the class option. And yes, we can add more auxiliary functions for authentication--as you mentioned--, verification,  and for checking the accessibility to API/database. Moreover,  such steps are done internally when to you call the read_from_* functions, but we can modularize them as separate functions.

Reply:
The idea of what I'm proposing is that users would not have to repeat the same arguments over and over if they are trying to get different pieces of data from the same source with successive `read_from_()` calls.

Reply:
We had this discussion internally here and we decided to make the 2 implementations and compare them.

My only reserves about this that, in `read_from_*`, we will have a long list of arguments (including both username, password, url, database specific arguments and api-specific arguments). 

The arguments were made in previous PRs that the list of arguments are to long and we need to act on this.

Reply:
Another option will consist in making the credential class such that it returns a connexion object as described below:

* It will have an `authenticate()` method that returns a connexion object. This method will be called internally by the class constructor.
* It's print() method will display the user's credential details and a string to tell whether the connexion was successfully established or not.

That way, the credential object (can be called authentication object!!!) is created once and will be used in any call of the `read_from_*` function, until the connection expires.